segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -136,52 +139,187 @@ public void initialize(long epoch) {
+
+        try {
+            return Retry.withExpBackoff(1, 5, MAX_ATTEMPT_COUNT)
+                        .retryingOn(FileNotFoundException.class)
+                        .throwingOn(IOException.class)
+                        .run(() -> {
+                            AtomicInteger totalBytesRead = new AtomicInteger();
+                            readInternal(handle, totalBytesRead, buffer, offset, bufferOffset, length);
+                            HDFSMetrics.READ_LATENCY.reportSuccessEvent(timer.getElapsed());
+                            HDFSMetrics.READ_BYTES.add(totalBytesRead.get());
+                            LoggerHelpers.traceLeave(log, "read", traceId, handle, offset, totalBytesRead);
+                            return totalBytesRead.get();
+                        });
+        } catch (IOException e) {
+            HDFSExceptionHelpers.throwException(handle.getSegmentName(), e);
+        } catch (RetriesExhaustedException e) {
+            HDFSExceptionHelpers.throwException(handle.getSegmentName(), e.getCause());
+        }
+        return -1;
     }
 
     @Override
-    public void write(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentException {
-        run(new WriteOperation(asWritableHandle(handle), offset, data, length, this.context));
+    public SegmentHandle openRead(String streamSegmentName) throws StreamSegmentException {
+        try {
+            //Ensure that file exists
+            findStatusForSegment(streamSegmentName, true);
+            return HDFSSegmentHandle.read(streamSegmentName, null);
+        } catch (IOException e) {
+            HDFSExceptionHelpers.throwException(streamSegmentName, e);
+            return null;
+        }
     }
 
     @Override
     public void seal(SegmentHandle handle) throws StreamSegmentException {
-        run(new SealOperation(asWritableHandle(handle), this.context));
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle);
+        handle = asWritableHandle(handle);
+        try {
+            FileStatus lastHandleFile = findStatusForSegment(handle.getSegmentName(), true);
+
+            if (!isSealed(lastHandleFile.getPath())) {
+                if (getEpoc(lastHandleFile) > this.context.epoch) {
+                    throw new StorageNotPrimaryException(handle.getSegmentName());
+                }
+                makeReadOnly(lastHandleFile);
+                Path sealedPath = getSealedFilePath(handle.getSegmentName());
+                this.context.fileSystem.rename(lastHandleFile.getPath(), sealedPath);
+            }
+        } catch (IOException e) {
+            HDFSExceptionHelpers.throwException(handle.getSegmentName(), e);
+        }
+
+        LoggerHelpers.traceLeave(log, "seal", traceId, handle);
     }
 
     @Override
     public void unseal(SegmentHandle handle) throws StreamSegmentException {
-        run(new UnsealOperation(asReadableHandle(handle), this.context));
+        try {
```

Comment:
Add trace enter/leave

Reply:
and in any other publicly visible method

Reply:
Fixed.